packages/utils/src/dal/mikro-orm/mikro-orm-repository.ts

```diff
@@ -420,6 +435,329 @@ export function mikroOrmBaseRepositoryFactory<T extends object = object>(
+        return this.getEntityWithId(manager, relation.type, normalizedItem)
+      })
+
+      if (relation.reference === ReferenceType.MANY_TO_MANY) {
+        const currentPivotColumn = relation.inverseJoinColumns[0]
+        const parentPivotColumn = relation.joinColumns[0]
+
+        if (!normalizedData.length) {
+          await manager.nativeDelete(relation.pivotEntity, {
+            [parentPivotColumn]: (data as any).id,
+          })
+
+          return normalizedData
+        }
+
+        await this.upsertMany_(manager, relation.type, normalizedData)
+
+        const pivotData = normalizedData.map((currModel) => {
+          return {
+            [parentPivotColumn]: (data as any).id,
+            [currentPivotColumn]: currModel.id,
+          }
+        })
+
+        const qb = manager.qb(relation.pivotEntity)
+        await qb.insert(pivotData).onConflict().ignore().execute()
+
+        await manager.nativeDelete(relation.pivotEntity, {
+          [parentPivotColumn]: (data as any).id,
+          [currentPivotColumn]: {
+            $nin: pivotData.map((d) => d[currentPivotColumn]),
+          },
+        })
+
+        return normalizedData
+      }
+
+      if (relation.reference === ReferenceType.ONE_TO_MANY) {
+        const joinColumns =
+          relation.targetMeta?.properties[relation.mappedBy]?.joinColumns
+
+        const joinColumnsConstraints = {}
+        joinColumns?.forEach((joinColumn, index) => {
+          const referencedColumnName = relation.referencedColumnNames[index]
+          joinColumnsConstraints[joinColumn] = data[referencedColumnName]
+        })
+
+        if (normalizedData.length) {
+          normalizedData.forEach((normalizedDataItem: any) => {
+            Object.assign(normalizedDataItem, {
+              ...joinColumnsConstraints,
+            })
+          })
+
+          await this.upsertMany_(manager, relation.type, normalizedData)
+        }
+
+        await manager.nativeDelete(relation.type, {
+          ...joinColumnsConstraints,
+          id: { $nin: normalizedData.map((d: any) => d.id) },
+        })
+
+        return normalizedData
+      }
+
+      return normalizedData
+    }
+
+    protected handleRelationAssignment_(
+      manager: SqlEntityManager,
+      upsertConfig: UpsertWithReplaceConfig<T>,
+      allRelationsToHandle: string[],
+      relation: EntityProperty<any>,
+      entryCopy: T,
+      upsertsPerType: Record<string, any[]>
+    ) {
+      if (
+        upsertConfig.relationsToUpsert?.includes(relation.name as keyof T) &&
+        relation.reference === ReferenceType.MANY_TO_ONE
+      ) {
+        // If it is undefined (not passed) then just return
+        if (entryCopy[relation.name] === undefined) {
+          return
+        }
+
+        // If null is passed (to be unset) then we need to set that on the join column
+        if (entryCopy[relation.name] === null) {
+          delete entryCopy[relation.name]
+          entryCopy[relation.joinColumns[0]] = null
```

Comment:
for this iteration, handling multiple primary/foreign keys have been punt to later

Reply:
@riqwan for now we only support ID-based keys in the function, so there will only be 1 (or it will fail otherwise). Adding support for composite key references will be a lot more work with very little benefit for now IMO